src/main/java/com/linkedin/kmf/services/ConsumeService.java

```diff
@@ -7,157 +7,148 @@
       _consumeThread = new Thread(() -> {
         try {
           consume();
         } catch (Exception e) {
-          LOG.error(_name + "/ConsumeService failed", e);
+          LOG.error(name + "/ConsumeService failed", e);
         }
-      }, _name + " consume-service");
+      }, name + " consume-service");
       _consumeThread.setDaemon(true);
-    });
+    }).get();
   }
 
   private void consume() throws Exception {
-    // Delay 1 second to reduce the chance that consumer creates topic before TopicManagementService
+    /* Delay 1 second to reduce the chance that consumer creates topic before TopicManagementService */
     Thread.sleep(1000);
 
     Map<Integer, Long> nextIndexes = new HashMap<>();
 
     while (_running.get()) {
       BaseConsumerRecord record;
       try {
-        record = _consumer.receive();
+        record = _baseConsumer.receive();
       } catch (Exception e) {
         _sensors._consumeError.record();
         LOG.warn(_name + "/ConsumeService failed to receive record", e);
-        // Avoid busy while loop
+        /* Avoid busy while loop */
         Thread.sleep(CONSUME_THREAD_SLEEP_MS);
         continue;
       }
 
-      if (record == null)
-        continue;
+      if (record == null) continue;
 
       GenericRecord avroRecord = Utils.genericRecordFromJson(record.value());
       if (avroRecord == null) {
         _sensors._consumeError.record();
         continue;
       }
       int partition = record.partition();
+
+      /* Commit availability and commit latency service */
+      try {
+        TopicPartition topicPartition = new TopicPartition(_topic, partition);
+        final AtomicReference<Exception> offsetCommitIssues = new AtomicReference<>(null);
+        OffsetAndMetadata offsetAndMetadata;
+        /* Call commitAsync, wait for a NON-NULL return value (see https://issues.apache.org/jira/browse/KAFKA-6183) */
+        OffsetCommitCallback commitCallback = (topicPartitionOffsetAndMetadataMap, exception) -> {
+          if (exception != null) {
+            offsetCommitIssues.set(exception);
+          }
+          _commitAvailabilityMetrics._offsetsCommitted.record();
+        };
+
+        long currTimeMillis = System.currentTimeMillis();
+
+        /* 5 seconds consumer offset commit interval. */
+        long timeDiffMillis = 5000;
+        if (currTimeMillis - _baseConsumer.lastCommitted() > timeDiffMillis) {
+          if (_offsetsToCommit != null && !_offsetsToCommit.isEmpty()) {
+            _baseConsumer.commitAsync(_offsetsToCommit, commitCallback);
+          } else {
+            _baseConsumer.commitAsync(commitCallback);
+          }
+
+          /* Record the current time for the committed consumer offset */
+          _baseConsumer.updateLastCommit();
+        }
+
+        offsetAndMetadata = _baseConsumer.committed(topicPartition);
+        if (offsetAndMetadata != null) {
+          _offsetsToCommit.put(topicPartition, offsetAndMetadata);
+        }
```

Comment:
I'm still confused about why this is needed and how the offsets for the consumer will not get stuck since it would always be committing the value it was committing before.

Reply:
Correct. Updated.